File: adventofcode/2022/day02/day02.hs
data Move = Rock | Paper | Scissor
  deriving (Show)

parseMove :: String -> Move
parseMove "A" = Rock
parseMove "X" = Rock
parseMove "B" = Paper
parseMove "Y" = Paper
parseMove "C" = Scissor
parseMove "Z" = Scissor

data Outcome = Lose | Draw | Win
  deriving (Show)

parseOutcome :: String -> Outcome
parseOutcome "X" = Lose
parseOutcome "Y" = Draw
parseOutcome "Z" = Win

moveScore :: Move -> Int
moveScore Rock = 1
moveScore Paper = 2
moveScore Scissor = 3

outcomeScore :: Outcome -> Int
outcomeScore Lose = 0
outcomeScore Draw = 3
outcomeScore Win = 6

outcome :: Move -> Move -> Outcome
outcome Rock Rock = Draw
outcome Rock Scissor = Lose
outcome Rock Paper = Win
outcome Paper Paper = Draw
outcome Paper Rock = Lose
outcome Paper Scissor = Win
outcome Scissor Scissor = Draw
outcome Scissor Paper = Lose
outcome Scissor Rock = Win

pickMove :: Move -> Outcome -> Move
pickMove Rock Draw = Rock
pickMove Rock Win = Paper
pickMove Rock Lose = Scissor
pickMove Paper Draw = Paper
pickMove Paper Win = Scissor
pickMove Paper Lose = Rock
pickMove Scissor Draw = Scissor
pickMove Scissor Win = Rock
pickMove Scissor Lose = Paper

roundScore :: Move -> Move -> Int
roundScore m1 m2 = moveScore m2 + outcomeScore out
  where out = outcome m1 m2

toTuple :: [a] -> (a, a)
toTuple [x, y] = (x, y)

toMoveOutcome :: (String, String) -> (Move, Move)
toMoveOutcome (x, y) = (m, pickMove m (parseOutcome y))
  where m = parseMove x

part1 :: [String] -> Int
part1 = sum . map (uncurry roundScore . toTuple . map parseMove . words)

part2 :: [String] -> Int
part2 = sum . map (uncurry roundScore . toMoveOutcome . toTuple . words)

main :: IO ()
main = interact $ unlines . map show . sequence [part1, part2] . lines

File: adventofcode/2022/day02/day02.zig
const std = @import("std");
const time = std.time;

const Move = enum {
    rock,
    paper,
    scissor,
};

fn parseMove(a: u8) Move {
    return switch (a) {
        'A', 'X' => Move.rock,
        'B', 'Y' => Move.paper,
        'C', 'Z' => Move.scissor,
        else => Move.rock,
    };
}

const Outcome = enum {
    win,
    draw,
    lose,
};

fn parseOutcome(o: u8) Outcome {
    return switch (o) {
        'X' => Outcome.lose,
        'Y' => Outcome.draw,
        'Z' => Outcome.win,
        else => Outcome.draw,
    };
}

fn parseMoveOutcome(a: u8, b: u8) [2]Move {
    const move = parseMove(a);
    return [2]Move{ move, pickMove(move, parseOutcome(b)) };
}

fn moveScore(move: Move) u8 {
    return switch (move) {
        .rock => 1,
        .paper => 2,
        .scissor => 3,
    };
}

fn outcomeScore(out: Outcome) u8 {
    return switch (out) {
        .lose => 0,
        .draw => 3,
        .win => 6,
    };
}

fn pickMove(move: Move, out: Outcome) Move {
    return switch (move) {
        .rock => {
            return switch (out) {
                .draw => Move.rock,
                .win => Move.paper,
                .lose => Move.scissor,
            };
        },
        .paper => {
            return switch (out) {
                .draw => Move.paper,
                .win => Move.scissor,
                .lose => Move.rock,
            };
        },
        .scissor => {
            return switch (out) {
                .draw => Move.scissor,
                .win => Move.rock,
                .lose => Move.paper,
            };
        },
    };
}

fn outcome(move1: Move, move2: Move) Outcome {
    return switch (move1) {
        .rock => {
            return switch (move2) {
                .rock => Outcome.draw,
                .scissor => Outcome.lose,
                .paper => Outcome.win,
            };
        },
        .paper => {
            return switch (move2) {
                .paper => Outcome.draw,
                .rock => Outcome.lose,
                .scissor => Outcome.win,
            };
        },
        .scissor => {
            return switch (move2) {
                .scissor => Outcome.draw,
                .paper => Outcome.lose,
                .rock => Outcome.win,
            };
        },
    };
}

fn roundScore(move1: Move, move2: Move) u32 {
    const out = outcome(move1, move2);
    return moveScore(move2) + outcomeScore(out);
}

pub fn main() !void {
    const s = time.microTimestamp();
    const stdin = std.io.getStdIn().reader();
    var p1: u32 = 0;
    var p2: u32 = 0;
    while (true) {
        var buf: [10]u8 = undefined;
        const line = try stdin.readUntilDelimiterOrEof(&buf, '\n');
        if (line == null) break;

        const lines = line.?;
        if (lines.len < 3) break;

        const move1 = parseMove(lines[0]);
        const move2 = parseMove(lines[2]);
        p1 += roundScore(move1, move2);

        const moves = parseMoveOutcome(lines[0], lines[2]);
        p2 += roundScore(moves[0], moves[1]);
    }
    std.debug.print("{d}\n", .{p1});
    std.debug.print("{d}\n", .{p2});
    const e = time.microTimestamp();
    std.debug.print("Time taken: {s}\n", .{std.fmt.fmtDurationSigned(e - s)});
}
